homeassistant/auth.py

```diff
@@ -0,0 +1,373 @@
+        return self._providers[auth_provider_key]
+
+
+class AuthStore:
+    """Stores authentication info.
+
+    Any mutation to an object should happen inside the auth store.
+
+    The auth store is lazy. It won't load the data from disk until a method is
+    called that needs it.
+    """
+
+    def __init__(self):
+        """Initialize the auth store."""
+        self.users = None
+
+    async def credentials_for_provider(self, provider_type, provider_id):
+        """Return credentials for specific auth provider type and id."""
+        if self.users is None:
+            await self.async_load()
+
+        result = []
+
+        for user in self.users:
+            for credentials in user.credentials:
+                if (credentials.auth_provider_type == provider_type and
+                        credentials.auth_provider_id == provider_id):
+                    result.append(credentials)
+
+        return result
+
+    async def async_get_user(self, user_id):
+        """Retrieve a user."""
+        if self.users is None:
+            await self.async_load()
+
+        for user in self.users:
+            if user.id == user_id:
+                return user
+
+        return None
+
+    async def async_get_or_create_user(self, credentials, auth_provider):
+        """Get or create a new user for given credentials.
+
+        If link_user is passed in, the credentials will be linked to the passed
+        in user if the credentials are new.
+        """
+        if self.users is None:
+            await self.async_load()
+
+        # New credentials, store in user
+        if credentials.id is None:
+            info = await auth_provider.async_user_meta_for_credentials(
+                credentials)
+            # Make owner and activate user if it's the first user.
+            if self.users:
+                is_owner = False
+                is_active = False
+            else:
+                is_owner = True
+                is_active = True
+
+            new_user = User(
+                is_owner=is_owner,
+                is_active=is_active,
+                name=info.get('name'),
+            )
+            self.users.append(new_user)
+            await self.async_link_user(new_user, credentials)
+            return new_user
+
+        for user in self.users:
+            for creds in user.credentials:
+                if (creds.auth_provider_type == credentials.auth_provider_type
+                        and creds.auth_provider_id == creds.auth_provider_id):
+                    return user
+
+        raise ValueError('We got credentials with ID but found no user')
+
+    async def async_link_user(self, user, credentials):
+        """Add credentials to an existing user."""
+        user.credentials.append(credentials)
+        await self.async_save()
+
+    async def async_load(self):
+        """Load the users."""
+        # TODO load from disk
+        self.users = []
```

Comment:
Maybe make this a dict of user ID and user?

Reply:
The problem is that if we're adding a user, it is not persisted yet. I guess we can change that by breaking up `async_save` and do the writes/assigning of IDs where needed

Reply:
Why did you want to wait with the ID generation? That isn't expensive is it?

Reply:
I was using it as an indicator that it wasn't saved yet.